lib/datalog/roletester/src/role_tester.rs

```diff
@@ -0,0 +1,220 @@
+    struct HasRole(u32, u32);
+    @input
+    struct HasTrait(u32, u32, u32);
+    @input
+    struct NodeHasLabel(u32, u32, u32);
+    @input
+    struct RoleAllowsNodeLabel(u32, u32, u32);
+    @input
+    struct RoleDeniesNodeLabel(u32, u32, u32);
+    @input
+    struct RoleAllowsLogin(u32, u32);
+    @input
+    struct RoleDeniesLogin(u32, u32);
+
+    // Intermediate rules
+    struct HasAllowNodeLabel(u32, u32, u32, u32);
+    struct HasDenyNodeLabel(u32, u32, u32, u32);
+    struct HasAllowRole(u32, u32, u32, u32);
+    struct HasDenyRole(u32, u32, u32);
+    struct HasDeniedLogin(u32, u32, u32);
+
+    // Output for IDB
+    @output
+    struct HasAccess(u32, u32, u32, u32);
+    @output
+    struct DenyAccess(u32, u32, u32, u32);
+    @output
+    struct DenyLogins(u32, u32, u32);
+
+    // Intermediate rules to help determine access
+    HasAllowNodeLabel(role, node, key, value) <- RoleAllowsNodeLabel(role, key, value), NodeHasLabel(node, key, value);
+    HasDenyNodeLabel(role, node, key, value) <- RoleDeniesNodeLabel(role, key, value), NodeHasLabel(node, key, value);
+    HasAllowRole(user, login, node, role) <- HasRole(user, role), HasAllowNodeLabel(role, node, _, _), RoleAllowsLogin(role, login),
+        !RoleDeniesLogin(role, login);
+    HasAllowRole(user, login, node, role) <- HasRole(user, role), HasAllowNodeLabel(role, node, _, _), HasTrait(user, LOGIN_TRAIT_HASH, login),
+        !RoleDeniesLogin(role, login), !RoleDeniesLogin(role, LOGIN_TRAIT_HASH);
+    HasDenyRole(user, node, role) <- HasRole(user, role), HasDenyNodeLabel(role, node, _, _);
+    HasDeniedLogin(user, login, role) <- HasRole(user, role), RoleDeniesLogin(role, login);
+    HasDeniedLogin(user, login, role) <- HasRole(user, role), HasTrait(user, LOGIN_TRAIT_HASH, login), RoleDeniesLogin(role, LOGIN_TRAIT_HASH);
+
+    // HasAccess rule determines each access for a specified user, login and node
+    HasAccess(user, login, node, role) <- HasAllowRole(user, login, node, role), !HasDenyRole(user, node, _), !HasDeniedLogin(user, login, _);
+    DenyAccess(user, login, node, role) <- HasDenyRole(user, node, role), HasTrait(user, LOGIN_TRAIT_HASH, login);
+    DenyAccess(user, login, node, role) <- HasDenyRole(user, node, role), HasAllowRole(user, login, node, _);
+
+    DenyLogins(user, login, role) <- HasDeniedLogin(user, login, role);
+}
+
+#[repr(C)]
+pub struct Status {
+    output_length: size_t,
+    error: i32
+}
+
+#[no_mangle]
+pub extern "C" fn process_access(
+    input: *mut c_uchar,
+    output: *mut c_uchar,
+    input_len: size_t,
+    output_len: size_t
+) -> *mut Status {
+    let mut runtime = Crepe::new();
+    let b = unsafe { slice::from_raw_parts_mut(input, input_len) };
+    let r = match types::Facts::decode(BytesMut::from(&b[..])) {
+        Ok(b) => b,
+        Err(e) => {
+            let err_bytes = e.to_string().into_bytes();
+            unsafe {
+                ptr::copy(&(*err_bytes)[0], output, err_bytes.len());
+            }
```

Comment:
Looks like this potentially performs an out of bounds write, since it writes the full length of `err_bytes`, but that length may be greater than `output_len`.

This is fixable by writing the minimum of the two lengths.   That being said, I think having the caller allocate output memory is an unnecessary complication.  Why not have go always allocate inputs and rust always allocate outputs?  That way you won't have to worry about sizing, and cleanup has a very straightforward model (go always deallocates inputs and rust always deallocates outputs).

Reply:
That's a good point. I'll have the buffer be returned in the struct.